Not able to Insert Neo4j Map Data type using the neo4j-spark connector

### Versions
like #315 

- Spark: 3.3.2
- Scala: 2.12
- Neo4j: neo4j-community-5.7.0
- Neo4j Connector:  neo4j-connector-apache-spark_2.12-4.0.1_for_spark_3.jar
- **Pyspark**
## Additional information
My neo4jDF printschema output

```markdown
root
 |-- id: string (nullable = true)
 |-- time: long (nullable = true)
 |-- table: map (nullable = true)
 |    |-- key: string
 |    |-- value: map (valueContainsNull = true)
 |    |    |-- key: string
 |    |    |-- value: string (valueContainsNull = true)
```

and get the error like this:
```markdown
Caused by: org.neo4j.driver.exceptions.ClientException: Unable to convert org.apache.spark.sql.catalyst.expressions.UnsafeMapData to Neo4j Value.
	at org.neo4j.driver.Values.value(Values.java:134)
	at org.neo4j.driver.Values.value(Values.java:297)
	at org.neo4j.driver.Values.value(Values.java:119)
	at org.neo4j.spark.util.Neo4jUtil$.convertFromSpark(Neo4jUtil.scala:210)
	at org.neo4j.spark.service.Neo4jWriteMappingStrategy.$anonfun$query$1(MappingService.scala:123)
	at org.neo4j.spark.service.Neo4jWriteMappingStrategy.$anonfun$query$1$adapted(MappingService.scala:121)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.immutable.Range.foreach(Range.scala:158)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at org.neo4j.spark.service.Neo4jWriteMappingStrategy.query(MappingService.scala:121)
	at org.neo4j.spark.service.Neo4jWriteMappingStrategy.node(MappingService.scala:30)
	at org.neo4j.spark.service.Neo4jWriteMappingStrategy.node(MappingService.scala:18)
	at org.neo4j.spark.service.MappingService.convert(MappingService.scala:230)
	at org.neo4j.spark.writer.BaseDataWriter.write(BaseDataWriter.scala:37)
	at org.neo4j.spark.writer.Neo4jDataWriter.write(Neo4jDataWriter.scala:9)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.$anonfun$run$1(WriteToDataSourceV2Exec.scala:442)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1538)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:480)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.$anonfun$writeWithV2$2(WriteToDataSourceV2Exec.scala:381)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:136)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more

```

--- a/common/src/main/scala/org/neo4j/spark/service/MappingService.scala
+++ b/common/src/main/scala/org/neo4j/spark/service/MappingService.scala
@@ -1,5 +1,7 @@
 package org.neo4j.spark.service
 
+import org.apache.spark.internal.Logging
+
 import java.util
 import java.util.function
 import java.util.function.BiConsumer
@@ -16,7 +18,8 @@ import scala.collection.mutable
 import org.neo4j.spark.util.Neo4jImplicits._
 
 class Neo4jWriteMappingStrategy(private val options: Neo4jOptions)
-  extends Neo4jMappingStrategy[InternalRow, java.util.Map[String, AnyRef]] {
+  extends Neo4jMappingStrategy[InternalRow, java.util.Map[String, AnyRef]]
+  with Logging {
 
   override def node(row: InternalRow, schema: StructType): java.util.Map[String, AnyRef] = {
     Validations.validate(ValidateSchemaOptions(options, schema))
@@ -122,10 +125,22 @@ class Neo4jWriteMappingStrategy(private val options: Neo4jOptions)
         val field = schema(i)
         val neo4jValue = Neo4jUtil.convertFromSpark(seq(i), field)
         neo4jValue match {
-          case map: MapValue => Neo4jUtil.flattenMap(map.asMap(), field.name)
-            .asScala
-            .map(t => (t._1, Values.value(t._2)))
-            .toSeq
+          case map: MapValue =>
+            val scalaMap = map.asMap().asScala.toMap
+            val dupKeys = scalaMap.flattenKeys(field.name)
+              .groupBy(identity)
+              .collect { case (x, List(_,_,_*)) => x }
+            if (dupKeys.nonEmpty) {
+              logWarning(
+                s"""
+                   |The field `${field.name}` which has a map {${scalaMap.mkString(", ")}}
+                   |contains the following duplicated keys: [${dupKeys.mkString(", ")}],
+                   |you will loose information stored in these keys
+                   |""".stripMargin)
+            }
+            scalaMap.flattenMap(field.name)
+              .mapValues(Values.value)
+              .toSeq
           case _ => Seq((field.name, neo4jValue))
         }
       })

--- a/common/src/main/scala/org/neo4j/spark/service/MappingService.scala
+++ b/common/src/main/scala/org/neo4j/spark/service/MappingService.scala
@@ -135,7 +135,7 @@ class Neo4jWriteMappingStrategy(private val options: Neo4jOptions)
                 s"""
                    |The field `${field.name}` which has a map {${scalaMap.mkString(", ")}}
                    |contains the following duplicated keys: [${dupKeys.mkString(", ")}],
-                   |you will loose information stored in these keys
+                   |you will lose some of the values associated with these duplicate keys
                    |""".stripMargin)
             }
             scalaMap.flattenMap(field.name)

--- a/common/src/main/scala/org/neo4j/spark/service/MappingService.scala
+++ b/common/src/main/scala/org/neo4j/spark/service/MappingService.scala
@@ -126,19 +126,8 @@ class Neo4jWriteMappingStrategy(private val options: Neo4jOptions)
         val neo4jValue = Neo4jUtil.convertFromSpark(seq(i), field)
         neo4jValue match {
           case map: MapValue =>
-            val scalaMap = map.asMap().asScala.toMap
-            val dupKeys = scalaMap.flattenKeys(field.name)
-              .groupBy(identity)
-              .collect { case (x, List(_,_,_*)) => x }
-            if (dupKeys.nonEmpty) {
-              logWarning(
-                s"""
-                   |The field `${field.name}` which has a map {${scalaMap.mkString(", ")}}
-                   |contains the following duplicated keys: [${dupKeys.mkString(", ")}],
-                   |you will lose some of the values associated with these duplicate keys
-                   |""".stripMargin)
-            }
-            scalaMap.flattenMap(field.name)
+            map.asMap().asScala.toMap
+              .flattenMap(field.name, options.schemaMetadata.mapGroupDuplicateKeys)
               .mapValues(Values.value)
               .toSeq
           case _ => Seq((field.name, neo4jValue))

--- a/common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala
+++ b/common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala
@@ -4,9 +4,11 @@ import org.apache.spark.sql.connector.expressions.Expression
 import org.apache.spark.sql.connector.expressions.aggregate.Aggregation
 import org.apache.spark.sql.sources.{And, EqualNullSafe, EqualTo, Filter, GreaterThan, GreaterThanOrEqual, In, IsNotNull, IsNull, LessThan, LessThanOrEqual, Not, Or, StringContains, StringEndsWith, StringStartsWith}
 import org.apache.spark.sql.types.{DataTypes, MapType, StructField, StructType}
+import org.json4s.jsonwritable
 import org.neo4j.driver.types.{Entity, Node, Relationship}
 import org.neo4j.spark.service.SchemaService
 
+import java.util
 import javax.lang.model.SourceVersion
 import scala.collection.JavaConverters._
 
@@ -196,6 +198,33 @@ object Neo4jImplicits {
 
   implicit class AggregationImplicit(aggregation: Aggregation) {
     def groupByCols(): Array[Expression] = ReflectionUtils.groupByCols(aggregation)
+  }
+
+  implicit class MapImplicit[K, V](map: Map[String, V]) {
+    def flattenMap(prefix: String = ""): Map[String, AnyRef] = {
+      map
+        .flatMap(t => {
+          val key: String = if (prefix != "") s"$prefix.${t._1}" else t._1
+          t._2 match {
+            case nestedMap: Map[String, _] => nestedMap.flattenMap(key).toSeq
+            case nestedMap: java.util.Map[String, _] => nestedMap.asScala.toMap.flattenMap(key).toSeq
+            case _ => Seq((key, t._2.asInstanceOf[AnyRef]))
+          }
+        })
+    }
+
+    def flattenKeys(prefix: String = ""): Seq[String] = {
+      map
+        .flatMap(t => {
+          val key: String = if (prefix != "") s"$prefix.${t._1}" else t._1
+          t._2 match {
+            case nestedMap: Map[String, _] => nestedMap.flattenKeys(key)
+            case nestedMap: java.util.Map[String, _] => nestedMap.asScala.toMap.flattenKeys(key)
+            case _ => Seq(key)
+          }
+        })
+        .toSeq
+    }
   }
 
 }

--- a/common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala
+++ b/common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala
@@ -1,14 +1,13 @@
 package org.neo4j.spark.util
 
+import org.apache.logging.log4j.{LogManager, Logger}
 import org.apache.spark.sql.connector.expressions.Expression
 import org.apache.spark.sql.connector.expressions.aggregate.Aggregation
 import org.apache.spark.sql.sources.{And, EqualNullSafe, EqualTo, Filter, GreaterThan, GreaterThanOrEqual, In, IsNotNull, IsNull, LessThan, LessThanOrEqual, Not, Or, StringContains, StringEndsWith, StringStartsWith}
 import org.apache.spark.sql.types.{DataTypes, MapType, StructField, StructType}
-import org.json4s.jsonwritable
 import org.neo4j.driver.types.{Entity, Node, Relationship}
 import org.neo4j.spark.service.SchemaService
 
-import java.util
 import javax.lang.model.SourceVersion
 import scala.collection.JavaConverters._
 
@@ -201,30 +200,34 @@ object Neo4jImplicits {
   }
 
   implicit class MapImplicit[K, V](map: Map[String, V]) {
-    def flattenMap(prefix: String = ""): Map[String, AnyRef] = {
-      map
-        .flatMap(t => {
-          val key: String = if (prefix != "") s"$prefix.${t._1}" else t._1
-          t._2 match {
-            case nestedMap: Map[String, _] => nestedMap.flattenMap(key).toSeq
-            case nestedMap: java.util.Map[String, _] => nestedMap.asScala.toMap.flattenMap(key).toSeq
-            case _ => Seq((key, t._2.asInstanceOf[AnyRef]))
-          }
-        })
-    }
 
-    def flattenKeys(prefix: String = ""): Seq[String] = {
-      map
-        .flatMap(t => {
-          val key: String = if (prefix != "") s"$prefix.${t._1}" else t._1
-          t._2 match {
-            case nestedMap: Map[String, _] => nestedMap.flattenKeys(key)
-            case nestedMap: java.util.Map[String, _] => nestedMap.asScala.toMap.flattenKeys(key)
-            case _ => Seq(key)
-          }
-        })
-        .toSeq
-    }
+    private def innerFlattenMap(map: Map[String, _], prefix: String): Seq[(String, AnyRef)] = map
+      .toSeq
+      .flatMap(t => {
+        val key: String = if (prefix != "") s"$prefix.${t._1}" else t._1
+        t._2 match {
+          case nestedMap: Map[String, _] => innerFlattenMap(nestedMap, key)
+          case nestedMap: java.util.Map[String, _] => innerFlattenMap(nestedMap.asScala.toMap, key)
+          case _ => Seq((key, t._2.asInstanceOf[AnyRef]))
+        }
+      })
+      .toList
+
+    def flattenMap(prefix: String = "", groupDuplicateKeys: Boolean = false): Map[String, AnyRef] = innerFlattenMap(map, prefix)
+      .groupBy(_._1)
+      .mapValues(seq => if (groupDuplicateKeys && seq.size > 1) seq.map(_._2).asJava else seq.last._2)
+      .toMap
+
+    def flattenKeys(prefix: String = ""): Seq[String] = map
+      .flatMap(t => {
+        val key: String = if (prefix != "") s"$prefix.${t._1}" else t._1
+        t._2 match {
+          case nestedMap: Map[String, _] => nestedMap.flattenKeys(key)
+          case nestedMap: java.util.Map[String, _] => nestedMap.asScala.toMap.flattenKeys(key)
+          case _ => Seq(key)
+        }
+      })
+      .toList
   }
 
 }

--- a/common/src/main/scala/org/neo4j/spark/util/Neo4jOptions.scala
+++ b/common/src/main/scala/org/neo4j/spark/util/Neo4jOptions.scala
@@ -54,7 +54,8 @@ class Neo4jOptions(private val options: java.util.Map[String, String]) extends S
 
   val schemaMetadata = Neo4jSchemaMetadata(getParameter(SCHEMA_FLATTEN_LIMIT, DEFAULT_SCHEMA_FLATTEN_LIMIT.toString).toInt,
     SchemaStrategy.withCaseInsensitiveName(getParameter(SCHEMA_STRATEGY, DEFAULT_SCHEMA_STRATEGY.toString).toUpperCase),
-    OptimizationType.withCaseInsensitiveName(getParameter(SCHEMA_OPTIMIZATION_TYPE, DEFAULT_OPTIMIZATION_TYPE.toString).toUpperCase))
+    OptimizationType.withCaseInsensitiveName(getParameter(SCHEMA_OPTIMIZATION_TYPE, DEFAULT_OPTIMIZATION_TYPE.toString).toUpperCase),
+    getParameter(SCHEMA_MAP_GROUP_DUPLICATE_KEYS, DEFAULT_MAP_GROUP_DUPLICATE_KEYS.toString).toBoolean)
 
   val query: Neo4jQueryOptions = (
     getParameter(QUERY.toString.toLowerCase),
@@ -216,7 +217,10 @@ case class Neo4jStreamingOptions(propertyName: String,
 
 case class Neo4jApocConfig(procedureConfigMap: Map[String, AnyRef])
 
-case class Neo4jSchemaMetadata(flattenLimit: Int, strategy: SchemaStrategy.Value, optimizationType: OptimizationType.Value)
+case class Neo4jSchemaMetadata(flattenLimit: Int,
+                               strategy: SchemaStrategy.Value,
+                               optimizationType: OptimizationType.Value,
+                               mapGroupDuplicateKeys: Boolean)
 case class Neo4jTransactionMetadata(retries: Int, failOnTransactionCodes: Set[String], batchSize: Int, retryTimeout: Long)
 
 case class Neo4jNodeMetadata(labels: Seq[String], nodeKeys: Map[String, String], nodeProps: Map[String, String])
@@ -370,6 +374,8 @@ object Neo4jOptions {
   val SCHEMA_STRATEGY = "schema.strategy"
   val SCHEMA_FLATTEN_LIMIT = "schema.flatten.limit"
   val SCHEMA_OPTIMIZATION_TYPE = "schema.optimization.type"
+  // map aggregation
+  val SCHEMA_MAP_GROUP_DUPLICATE_KEYS = "schema.map.group.duplicate.keys"
 
   // partitions
   val PARTITIONS = "partitions"
@@ -383,7 +389,6 @@ object Neo4jOptions {
 
   val BATCH_SIZE = "batch.size"
   val SUPPORTED_SAVE_MODES = Seq(SaveMode.Overwrite, SaveMode.ErrorIfExists, SaveMode.Append)
-  val SUPPORTED_SAVE_MODES_FOR_STREAMING = Seq(SaveMode.ErrorIfExists, SaveMode.Overwrite, SaveMode.Append)
 
   // Relationship Metadata
   val RELATIONSHIP_SOURCE_LABELS = s"${QueryType.RELATIONSHIP.toString.toLowerCase}.source.${QueryType.LABELS.toString.toLowerCase}"
@@ -444,6 +449,8 @@ object Neo4jOptions {
   // Default values optimizations for Aura please look at: https://aura.support.neo4j.com/hc/en-us/articles/1500002493281-Neo4j-Java-driver-settings-for-Aura
   val DEFAULT_CONNECTION_MAX_LIFETIME_MSECS = Duration.ofMinutes(8).toMillis
   val DEFAULT_CONNECTION_LIVENESS_CHECK_TIMEOUT_MSECS = Duration.ofMinutes(2).toMillis
+
+  val DEFAULT_MAP_GROUP_DUPLICATE_KEYS = false
 }
 
 class CaseInsensitiveEnumeration extends Enumeration {

--- a/common/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
+++ b/common/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
@@ -211,7 +211,7 @@ object Neo4jUtil {
       val map: Map[String, AnyRef] = (0 to unsafeMapData.numElements() - 1)
         .map(i => (unsafeMapData.keyArray().getUTF8String(i).toString, unsafeMapData.valueArray().get(i, mapType.valueType)))
         .toMap[String, AnyRef]
-        .mapValues(innerValue => convertFromSpark(innerValue))
+        .mapValues(innerValue => convertFromSpark(innerValue, StructField("", mapType.valueType, true)))
         .toMap[String, AnyRef]
       Values.value(map.asJava)
     }
@@ -220,13 +220,15 @@ object Neo4jUtil {
   }
 
   def flattenMap(map: java.util.Map[String, AnyRef],
-                 prefix: String = ""): java.util.Map[String, AnyRef] = map.asScala.flatMap(t => {
-    val key: String = if (prefix != "") s"${prefix.quote()}.${t._1.quote()}" else t._1.quote()
-    t._2 match {
-      case nestedMap: Map[String, AnyRef] => flattenMap(nestedMap.asJava, key).asScala.toSeq
-      case _ => Seq((key, t._2))
-    }
-  })
+                 prefix: String = ""): java.util.Map[String, AnyRef] = map.asScala
+    .flatMap(t => {
+      val key: String = if (prefix != "") s"${prefix}.${t._1}" else t._1.quote()
+      t._2 match {
+        case nestedMap: Map[String, AnyRef] => flattenMap(nestedMap.asJava, key).asScala.toSeq
+        case nestedMap: java.util.Map[String, AnyRef] => flattenMap(nestedMap, key).asScala.toSeq
+        case _ => Seq((key, t._2))
+      }
+    })
     .toMap
     .asJava
 

--- a/common/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
+++ b/common/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
@@ -219,19 +219,6 @@ object Neo4jUtil {
     case _ => Values.value(value)
   }
 
-  def flattenMap(map: java.util.Map[String, AnyRef],
-                 prefix: String = ""): java.util.Map[String, AnyRef] = map.asScala
-    .flatMap(t => {
-      val key: String = if (prefix != "") s"${prefix}.${t._1}" else t._1.quote()
-      t._2 match {
-        case nestedMap: Map[String, AnyRef] => flattenMap(nestedMap.asJava, key).asScala.toSeq
-        case nestedMap: java.util.Map[String, AnyRef] => flattenMap(nestedMap, key).asScala.toSeq
-        case _ => Seq((key, t._2))
-      }
-    })
-    .toMap
-    .asJava
-
   def isLong(str: String): Boolean = {
     if (str == null) {
       false

--- a/common/src/test/scala/org/neo4j/spark/util/Neo4jImplicitsTest.scala
+++ b/common/src/test/scala/org/neo4j/spark/util/Neo4jImplicitsTest.scala
@@ -5,8 +5,11 @@ import org.apache.spark.sql.connector.expressions.aggregate.{Aggregation, Sum}
 import org.apache.spark.sql.sources.{And, EqualTo}
 import org.apache.spark.sql.types.{DataTypes, StructField, StructType}
 import org.junit.Assert._
-import org.junit.Test
+import org.junit.{Assert, Test}
 import org.neo4j.spark.util.Neo4jImplicits._
+
+import scala.collection.JavaConverters.{mapAsJavaMapConverter, mapAsScalaMapConverter}
+import scala.collection.immutable.ListMap
 
 class Neo4jImplicitsTest {
 
@@ -140,5 +143,53 @@ class Neo4jImplicitsTest {
     val agg = new Aggregation(Array(new Sum(aggField, false)), Array(gbyField))
     assertEquals(1, agg.groupByCols().length)
     assertEquals("bar", agg.groupByCols()(0).describe())
+  }
+
+  @Test
+  def `should flatten the map`(): Unit = {
+    val input = Map(
+      "foo" -> "bar",
+      "key" -> Map(
+        "innerKey" -> Map("innerKey2" -> "value")
+      )
+    )
+    val expected = Map(
+      "foo" -> "bar",
+      "key.innerKey.innerKey2" -> "value"
+    )
+    val actual = input.flattenMap()
+    Assert.assertEquals(expected, actual)
+  }
+
+  @Test
+  def `should not handle collision`(): Unit = {
+    val input = ListMap(
+      "my" -> Map(
+        "inner" -> Map("key" -> 42424242),
+        "inner.key" -> 424242
+      ),
+      "my.inner" -> Map("key" -> 4242).asJava,
+      "my.inner.key" -> 42
+    )
+    val expected = Map(
+      "my.inner.key" -> 42
+    )
+    val actual = input.flattenMap()
+    Assert.assertEquals(expected, actual)
+  }
+
+  @Test
+  def `should show duplicate keys`(): Unit = {
+    val input = Map(
+      "my" -> Map(
+        "inner" -> Map("key" -> 42424242),
+        "inner.key" -> 424242
+      ),
+      "my.inner" -> Map("key" -> 4242).asJava,
+      "my.inner.key" -> 42
+    )
+    val expected = Seq("my.inner.key", "my.inner.key", "my.inner.key", "my.inner.key")
+    val actual = input.flattenKeys()
+    Assert.assertEquals(expected, actual)
   }
 }

--- a/common/src/test/scala/org/neo4j/spark/util/Neo4jImplicitsTest.scala
+++ b/common/src/test/scala/org/neo4j/spark/util/Neo4jImplicitsTest.scala
@@ -8,7 +8,7 @@ import org.junit.Assert._
 import org.junit.{Assert, Test}
 import org.neo4j.spark.util.Neo4jImplicits._
 
-import scala.collection.JavaConverters.{mapAsJavaMapConverter, mapAsScalaMapConverter}
+import scala.collection.JavaConverters.{mapAsJavaMapConverter, mapAsScalaMapConverter, seqAsJavaListConverter}
 import scala.collection.immutable.ListMap
 
 class Neo4jImplicitsTest {
@@ -175,6 +175,23 @@ class Neo4jImplicitsTest {
       "my.inner.key" -> 42
     )
     val actual = input.flattenMap()
+    Assert.assertEquals(expected, actual)
+  }
+
+  @Test
+  def `should handle collision by aggregating values`(): Unit = {
+    val input = ListMap(
+      "my" -> Map(
+        "inner" -> Map("key" -> 42424242),
+        "inner.key" -> 424242
+      ),
+      "my.inner" -> Map("key" -> 4242).asJava,
+      "my.inner.key" -> 42
+    )
+    val expected = Map(
+      "my.inner.key" -> Seq(42424242, 424242, 4242, 42).asJava
+    )
+    val actual = input.flattenMap(groupDuplicateKeys = true)
     Assert.assertEquals(expected, actual)
   }
 

--- a/common/src/test/scala/org/neo4j/spark/util/Neo4jUtilTest.scala
+++ b/common/src/test/scala/org/neo4j/spark/util/Neo4jUtilTest.scala
@@ -11,19 +11,4 @@ class Neo4jUtilTest {
     Neo4jUtil.closeSafety(null)
   }
 
-  @Test
-  def testFlattenMap(): Unit = {
-    val input: java.util.Map[String, AnyRef] = Map(
-      "foo" -> "bar",
-      "key" -> Map(
-        "innerKey" -> Map("innnerKey2" -> "value").asJava
-      ).asJava
-    ).asJava
-    val expected = Map(
-      "foo" -> "bar",
-      "key.innerKey.innnerKey2" -> "value"
-    ).asJava
-    val actual = Neo4jUtil.flattenMap(input)
-    Assert.assertEquals(expected, actual)
-  }
 }

--- a/common/src/test/scala/org/neo4j/spark/util/Neo4jUtilTest.scala
+++ b/common/src/test/scala/org/neo4j/spark/util/Neo4jUtilTest.scala
@@ -1,11 +1,29 @@
 package org.neo4j.spark.util
 
-import org.junit.Test
+import org.junit.{Assert, Test}
+
+import scala.collection.JavaConverters.mapAsJavaMapConverter
 
 class Neo4jUtilTest {
 
   @Test
   def testSafetyCloseShouldNotFailWithNull(): Unit = {
     Neo4jUtil.closeSafety(null)
+  }
+
+  @Test
+  def testFlattenMap(): Unit = {
+    val input: java.util.Map[String, AnyRef] = Map(
+      "foo" -> "bar",
+      "key" -> Map(
+        "innerKey" -> Map("innnerKey2" -> "value").asJava
+      ).asJava
+    ).asJava
+    val expected = Map(
+      "foo" -> "bar",
+      "key.innerKey.innnerKey2" -> "value"
+    ).asJava
+    val actual = Neo4jUtil.flattenMap(input)
+    Assert.assertEquals(expected, actual)
   }
 }

--- a/doc/docs/modules/ROOT/pages/writing.adoc
+++ b/doc/docs/modules/ROOT/pages/writing.adoc
@@ -708,3 +708,76 @@ CREATE (n:Person{fullName: event.name + ' ' + event.surname, age: scriptResult[0
 
 `scriptResult` is the result from the last query contained within the `script` options
 that is `RETURN 36 AS age;`
+
+== Note about columns with Map type
+
+When a Dataframe column is a map, what we do internally is to flatten the map as Neo4j does not support this type for graph entity properties; so for a Spark job like this:
+
+[source,scala]
+----
+val data = Seq(
+  ("Foo", 1, Map("inner" -> Map("key" -> "innerValue"))),
+  ("Bar", 2, Map("inner" -> Map("key" -> "innerValue1"))),
+).toDF("id", "time", "table")
+
+data.write
+  .mode(SaveMode.Append)
+  .format(classOf[DataSource].getName)
+  .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+  .option("labels", ":MyNodeWithMapFlattend")
+  .save()
+----
+
+In Neo4j for the nodes with label `MyNodeWithMapFlattend` you'll find this information stored:
+
+----
+MyNodeWithMapFlattend {
+    id: 'Foo',
+    time: 1,
+    `table.inner.key`: 'innerValue'
+}
+MyNodeWithMapFlattend {
+    id: 'Bar',
+    time: 1,
+    `table.inner.key`: 'innerValue1'
+}
+----
+
+Now could happen that you have a Spark job like
+
+[source,scala]
+----
+val data = Seq(
+  ("Foo", 1, Map("key.inner" -> Map("key" -> "innerValue"), "key" -> Map("inner.key" -> "value"))),
+  ("Bar", 1, Map("key.inner" -> Map("key" -> "innerValue1"), "key" -> Map("inner.key" -> "value1"))),
+).toDF("id", "time", "table")
+data.write
+  .mode(SaveMode.Append)
+  .format(classOf[DataSource].getName)
+  .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+  .option("labels", ":MyNodeWithMapFlattend")
+  .save()
+----
+
+which leads to have duplicates keys while we flatten it, the Neo4j Spark connector informs you with a WARN message like this;
+
+----
+The field `table` which has a map {key.inner -> {key=innerValue}, key -> {inner.key=value}}
+contains the following duplicated keys: [table.key.inner.key],
+you will loose information stored in these keys
+----
+
+Because the information that we'll store into Neo4j will be this:
+
+----
+MyNodeWithMapFlattend {
+    id: 'Foo',
+    time: 1,
+    `table.key.inner.key`: 'innerValue'
+}
+MyNodeWithMapFlattend {
+    id: 'Bar',
+    time: 1,
+    `table.key.inner.key`: 'innerValue1'
+}
+----

--- a/doc/docs/modules/ROOT/pages/writing.adoc
+++ b/doc/docs/modules/ROOT/pages/writing.adoc
@@ -743,7 +743,7 @@ MyNodeWithFlattenedMap {
 }
 ----
 
-Now could  into problematic situations like the following one:
+Now you could fall into problematic situations like the following one:
 
 [source,scala]
 ----
@@ -759,13 +759,7 @@ data.write
   .save()
 ----
 
-since the resulting flattened keys are duplicated, the Neo4j Spark will pick one of the associated value in a non-deterministic way and inform you with a `WARN` message like the following one:
-
-----
-The field `table` which has a map {key.inner -> {key=innerValue}, key -> {inner.key=value}}
-contains the following duplicated keys: [table.key.inner.key],
-you will lose some of the values associated with these duplicate keys
-----
+since the resulting flattened keys are duplicated, the Neo4j Spark will pick one of the associated value in a non-deterministic way.
 
 Because the information that we'll store into Neo4j will be this (consider that the order is not guaranteed):
 
@@ -781,3 +775,39 @@ MyNodeWithFlattenedMap {
     `table.key.inner.key`: 'innerValue1' // but it could be `value1` as the order is not guaranteed
 }
 ----
+
+=== Group duplicated keys to array of values
+
+You can use the option `schema.map.group.duplicate.keys` to avoid this problem. The connector will group all the values with the same keys into an array. The default value for the option is `false`.
+In a scenario like this:
+
+[source,scala]
+----
+val data = Seq(
+  ("Foo", 1, Map("key.inner" -> Map("key" -> "innerValue"), "key" -> Map("inner.key" -> "value"))),
+  ("Bar", 1, Map("key.inner" -> Map("key" -> "innerValue1"), "key" -> Map("inner.key" -> "value1"))),
+).toDF("id", "time", "table")
+data.write
+  .mode(SaveMode.Append)
+  .format(classOf[DataSource].getName)
+  .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+  .option("labels", ":MyNodeWithFlattenedMap")
+  .option("schema.map.group.duplicate.keys", true)
+  .save()
+----
+
+the output would be:
+
+----
+MyNodeWithFlattenedMap {
+    id: 'Foo',
+    time: 1,
+    `table.key.inner.key`: ['innerValue', 'value'] // the order is not guaranteed
+}
+MyNodeWithFlattenedMap {
+    id: 'Bar',
+    time: 1,
+    `table.key.inner.key`: ['innerValue1', 'value1'] // the order is not guaranteed
+}
+----
+

--- a/doc/docs/modules/ROOT/pages/writing.adoc
+++ b/doc/docs/modules/ROOT/pages/writing.adoc
@@ -724,26 +724,26 @@ data.write
   .mode(SaveMode.Append)
   .format(classOf[DataSource].getName)
   .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
-  .option("labels", ":MyNodeWithMapFlattend")
+  .option("labels", ":MyNodeWithFlattenedMap")
   .save()
 ----
 
-In Neo4j for the nodes with label `MyNodeWithMapFlattend` you'll find this information stored:
+In Neo4j for the nodes with label `MyNodeWithFlattenedMap` you'll find this information stored:
 
 ----
-MyNodeWithMapFlattend {
+MyNodeWithFlattenedMap {
     id: 'Foo',
     time: 1,
     `table.inner.key`: 'innerValue'
 }
-MyNodeWithMapFlattend {
+MyNodeWithFlattenedMap {
     id: 'Bar',
     time: 1,
     `table.inner.key`: 'innerValue1'
 }
 ----
 
-Now could happen that you have a Spark job like
+Now could  into problematic situations like the following one:
 
 [source,scala]
 ----
@@ -755,29 +755,29 @@ data.write
   .mode(SaveMode.Append)
   .format(classOf[DataSource].getName)
   .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
-  .option("labels", ":MyNodeWithMapFlattend")
+  .option("labels", ":MyNodeWithFlattenedMap")
   .save()
 ----
 
-which leads to have duplicates keys while we flatten it, the Neo4j Spark connector informs you with a WARN message like this;
+since the resulting flattened keys are duplicated, the Neo4j Spark will pick one of the associated value in a non-deterministic way and inform you with a `WARN` message like the following one:
 
 ----
 The field `table` which has a map {key.inner -> {key=innerValue}, key -> {inner.key=value}}
 contains the following duplicated keys: [table.key.inner.key],
-you will loose information stored in these keys
+you will lose some of the values associated with these duplicate keys
 ----
 
-Because the information that we'll store into Neo4j will be this:
+Because the information that we'll store into Neo4j will be this (consider that the order is not guaranteed):
 
 ----
-MyNodeWithMapFlattend {
+MyNodeWithFlattenedMap {
     id: 'Foo',
     time: 1,
-    `table.key.inner.key`: 'innerValue'
+    `table.key.inner.key`: 'innerValue' // but it could be `value` as the order is not guaranteed
 }
-MyNodeWithMapFlattend {
+MyNodeWithFlattenedMap {
     id: 'Bar',
     time: 1,
-    `table.key.inner.key`: 'innerValue1'
+    `table.key.inner.key`: 'innerValue1' // but it could be `value1` as the order is not guaranteed
 }
 ----

--- a/spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
+++ b/spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
@@ -15,6 +15,7 @@ import org.neo4j.driver.{Result, Transaction, TransactionWork}
 import org.neo4j.spark.util.Neo4jOptions
 
 import scala.collection.JavaConverters._
+import scala.collection.immutable.ListMap
 import scala.util.Random
 
 abstract class Neo4jType(`type`: String)
@@ -1436,8 +1437,8 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
   @Test
   def shouldFix502WithCollisions(): Unit = {
     val data = Seq(
-      ("Foo", 1, Map("key.inner" -> Map("key" -> "innerValue"), "key" -> Map("inner.key" -> "value"))),
-      ("Bar", 1, Map("key.inner" -> Map("key" -> "innerValue1"), "key" -> Map("inner.key" -> "value1"))),
+      ("Foo", 1, ListMap("key.inner" -> Map("key" -> "innerValue"), "key" -> Map("inner.key" -> "value"))),
+      ("Bar", 1, ListMap("key.inner" -> Map("key" -> "innerValue1"), "key" -> Map("inner.key" -> "value1"))),
     ).toDF("id", "time", "table")
     data.write
       .mode(SaveMode.Append)
@@ -1454,7 +1455,35 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
         |)
         |RETURN count(n)
         |""".stripMargin)
-      .peek()
+      .single()
+      .get(0)
+      .asLong()
+    junit.Assert.assertEquals(2L, count)
+  }
+
+  @Test
+  def shouldFix502WithCollisionsAndAggregateValues(): Unit = {
+    val data = Seq(
+      ("Foo", 1, ListMap("key.inner" -> Map("key" -> "innerValue"), "key" -> Map("inner.key" -> "value"))),
+      ("Bar", 1, ListMap("key.inner" -> Map("key" -> "innerValue1"), "key" -> Map("inner.key" -> "value1"))),
+    ).toDF("id", "time", "table")
+    data.write
+      .mode(SaveMode.Append)
+      .format(classOf[DataSource].getName)
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .option("labels", ":MyNodeWithMapFlattend")
+      .option("schema.map.group.duplicate.keys", true)
+      .save()
+    val count: Long = SparkConnectorScalaSuiteIT.session().run(
+      """
+        |MATCH (n:MyNodeWithMapFlattend)
+        |WHERE (
+        | properties(n) = {id: 'Foo', time: 1, `table.key.inner.key`: ['innerValue', 'value']}
+        | OR properties(n) = {id: 'Bar', time: 1, `table.key.inner.key`: ['innerValue1', 'value1']}
+        |)
+        |RETURN count(n)
+        |""".stripMargin)
+      .single()
       .get(0)
       .asLong()
     junit.Assert.assertEquals(2L, count)

--- a/spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
+++ b/spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
@@ -4,6 +4,7 @@ import java.time.{LocalTime, OffsetTime, ZoneOffset}
 import org.apache.commons.lang3.exception.ExceptionUtils
 import org.apache.spark.SparkException
 import org.apache.spark.sql.{DataFrame, SaveMode, SparkSession}
+import org.junit
 import org.junit.Assert._
 import org.junit.{Ignore, Test}
 import org.neo4j.driver.internal.types.InternalTypeSystem
@@ -1361,8 +1362,8 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
   @Test
   def shouldWriteComplexDF(): Unit = {
     val data = Seq(
-      ("Cuba Gooding Jr.", 1, "2022-06-07 00:00:00", Seq(Map("product_id" -> 1, "quantity" -> 2, "product_id" -> 2, "quantity" -> 4))),
-      ("Tom Hanks", 2, "2022-07-07 00:00:00", Seq(Map("product_id" -> 11, "quantity" -> 2, "product_id" -> 22, "quantity" -> 4)))
+      ("Cuba Gooding Jr.", 1, "2022-06-07 00:00:00", Seq(Map("product_id" -> 1, "quantity" -> 2), Map("product_id" -> 2, "quantity" -> 4))),
+      ("Tom Hanks", 2, "2022-07-07 00:00:00", Seq(Map("product_id" -> 11, "quantity" -> 2), Map("product_id" -> 22, "quantity" -> 4)))
     ).toDF("actor_name", "order_id", "order_date", "products")
     data.write
       .mode(SaveMode.Overwrite)
@@ -1383,6 +1384,8 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
     val actual = SparkConnectorScalaSuiteIT.session().run(
       """
         |MATCH (p:Person)-[cr:CREATED]->(o:Order)-[co:CONTAINS]->(pr:Product)
+        |WITH p, pr, o, co
+        |ORDER BY p.name, pr.id
         |RETURN p.name AS name, o.id AS order, collect({id: pr.id, quantity: co.quantityOrdered}) AS products
         |""".stripMargin)
       .list()
@@ -1391,9 +1394,42 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
       .toSet
       .asJava
     val expected = Set(
-      Map("name" -> "Cuba Gooding Jr.", "order" -> 1L, "products" -> List(Map("quantity" -> 4L, "id" -> 2L).asJava).asJava).asJava,
-      Map("name" -> "Tom Hanks", "order" -> 2L, "products" -> List(Map("quantity" -> 4L, "id" -> 22L).asJava).asJava).asJava
+      Map("name" -> "Cuba Gooding Jr.", "order" -> 1L, "products" -> List(
+        Map("id" -> 1L, "quantity" -> 2L).asJava,
+        Map("id" -> 2L, "quantity" -> 4L).asJava
+      ).asJava).asJava,
+      Map("name" -> "Tom Hanks", "order" -> 2L, "products" -> List(
+        Map("id" -> 11L, "quantity" -> 2L).asJava,
+        Map("id" -> 22L, "quantity" -> 4L).asJava
+      ).asJava).asJava
     ).asJava
     assertEquals(expected, actual)
+  }
+
+  @Test
+  def shouldFix502(): Unit = {
+    val data = Seq(
+      ("Foo", 1, Map("key" -> Map("innerKey" -> "value"))),
+      ("Bar", 1, Map("key" -> Map("innerKey" -> "value1"))),
+    ).toDF("id", "time", "table")
+    data.write
+      .mode(SaveMode.Append)
+      .format(classOf[DataSource].getName)
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .option("labels", ":MyNodeWithMapFlattend")
+      .save()
+    val count: Long = SparkConnectorScalaSuiteIT.session().run(
+      """
+        |MATCH (n:MyNodeWithMapFlattend)
+        |WHERE (
+        | properties(n) = {id: 'Foo', time: 1, `table.key.innerKey`: 'value'}
+        | OR properties(n) = {id: 'Bar', time: 1, `table.key.innerKey`: 'value1'}
+        |)
+        |RETURN count(n)
+        |""".stripMargin)
+      .peek()
+      .get(0)
+      .asLong()
+    junit.Assert.assertEquals(2L, count)
   }
 }

--- a/spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
+++ b/spark-3/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
@@ -1409,8 +1409,8 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
   @Test
   def shouldFix502(): Unit = {
     val data = Seq(
-      ("Foo", 1, Map("key" -> Map("innerKey" -> "value"))),
-      ("Bar", 1, Map("key" -> Map("innerKey" -> "value1"))),
+      ("Foo", 1, Map("inner" -> Map("key" -> "innerValue"))),
+      ("Bar", 1, Map("inner" -> Map("key" -> "innerValue1"))),
     ).toDF("id", "time", "table")
     data.write
       .mode(SaveMode.Append)
@@ -1422,8 +1422,35 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
       """
         |MATCH (n:MyNodeWithMapFlattend)
         |WHERE (
-        | properties(n) = {id: 'Foo', time: 1, `table.key.innerKey`: 'value'}
-        | OR properties(n) = {id: 'Bar', time: 1, `table.key.innerKey`: 'value1'}
+        | properties(n) = {id: 'Foo', time: 1, `table.inner.key`: 'innerValue'}
+        | OR properties(n) = {id: 'Bar', time: 1, `table.inner.key`: 'innerValue1'}
+        |)
+        |RETURN count(n)
+        |""".stripMargin)
+      .peek()
+      .get(0)
+      .asLong()
+    junit.Assert.assertEquals(2L, count)
+  }
+
+  @Test
+  def shouldFix502WithCollisions(): Unit = {
+    val data = Seq(
+      ("Foo", 1, Map("key.inner" -> Map("key" -> "innerValue"), "key" -> Map("inner.key" -> "value"))),
+      ("Bar", 1, Map("key.inner" -> Map("key" -> "innerValue1"), "key" -> Map("inner.key" -> "value1"))),
+    ).toDF("id", "time", "table")
+    data.write
+      .mode(SaveMode.Append)
+      .format(classOf[DataSource].getName)
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .option("labels", ":MyNodeWithMapFlattend")
+      .save()
+    val count: Long = SparkConnectorScalaSuiteIT.session().run(
+      """
+        |MATCH (n:MyNodeWithMapFlattend)
+        |WHERE (
+        | properties(n) = {id: 'Foo', time: 1, `table.key.inner.key`: 'value'}
+        | OR properties(n) = {id: 'Bar', time: 1, `table.key.inner.key`: 'value1'}
         |)
         |RETURN count(n)
         |""".stripMargin)

